### PR TITLE
External memory management

### PR DESCRIPTION
**Pull request by NickNaso**
Added utility function `AdjustExternalMemory` like discussed in the following issue:
[https://github.com/nodejs/node-addon-api/issues/260](https://github.com/nodejs/node-addon-api/issues/260)
I added the function `AdjustExternalMemory` with corresponding tests and documentation. Please give a review of my work.

### CODE REVIEW

**Comment by mhdawson** [comment]
Hope to find time to review tomorrow.

File: doc/error.md [hunk 1]
```diff
@@ -1,5 +1,114 @@
 # Error
 
-You are reading a draft of the next documentation and it's in continuos update so
-if you don't find what you need please refer to:
-[C++ wrapper classes for the ABI-stable C APIs for Node.js](https://nodejs.github.io/node-addon-api/)
\ No newline at end of file
+The **Error** class is a representation of the JavaScript Error object that is thrown
```
Comment (mhdawson) [inline doc/error.md]: Should this file be updated in this PR?

File: doc/error_handling.md [hunk 2]
```diff
@@ -1,5 +1,151 @@
 # Error handling
 
-You are reading a draft of the next documentation and it's in continuos update so
-if you don't find what you need please refer to:
-[C++ wrapper classes for the ABI-stable C APIs for Node.js](https://nodejs.github.io/node-addon-api/)
\ No newline at end of file
+Error handling represents one of the most important considerations when
```
Comment (mhdawson) [inline doc/error_handling.md]: Should this file b e updated in this PR?

File: doc/memory_management.md [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+more often than it would otherwise in an attempt to garbage collect the JavaScript
+objects that keep the externally allocated memory alive.
+
+## AdjustExternalMemory
+
+This function gives the JavaScript engine an indication of the amount of externally
+allocated memory that is kept alive by JavaScript objects.
+
+```cpp
+int64_t AdjustExternalMemory(napi_env env, int64_t change_in_bytes);
```
Comment (mhdawson) [inline doc/memory_management.md]: I think the env parameter should probably be a node-addon-api  Env as opposed to a napi_env.

File: doc/range_error.md [hunk 4]
```diff
@@ -0,0 +1,59 @@
+# RangeError
```
Comment (mhdawson) [inline doc/range_error.md]: Don't think this file should be changed in this PR.

**Comment by mhdawson** [comment]
Generally looks ok, except there are files that I don't think should be changed in this PR.  Otherwise only a few small comments.

**Comment by NickNaso** [comment]
@mhdawson I changed the `AdjustExternalMemory` function as requested. I wrote the documentation for this function in file `doc/memory_management`. Could you take a look? After your revision I will provide to remove the wrong commit.

File: napi-inl.h [hunk 5]
```diff
@@ -3074,6 +3074,17 @@ inline Value EscapableHandleScope::Escape(napi_value escapee) {
   return Value(_env, result);
 }
 
+////////////////////////////////////////////////////////////////////////////////
```
Comment (mhdawson) [inline napi-inl.h]: This seems to add just a method as opposed to a class with methods.  I think in earlier discussion it was either going to be a MemoryManagement class (or may Util class) with the method to adjust the memory.

Comment (NickNaso) [reply napi-inl.h]: It was my misunderstanding. Now I will provide to create a class that expose the method. Sorry.

**Comment by mhdawson** [comment]
This PR seems to change a lot of unrelated files.  Can you fix up that up or open a new PR as an alternative?

**Comment by NickNaso** [comment]
Hi @mhdawson please close this PR and later I will provide to open a new one.

**Comment by mhdawson** [comment]
Closing and will look for new PR.

**Pull request closed without merging**